Update the Interaction Regions layer tree after UI-side scrolling

etiennesegonzac opened the pull request:
#### 8f2972ef9f74c7f00d74077697c4028a1121773e
<pre>
Update the Interaction Regions layer tree after UI-side scrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=251696">https://bugs.webkit.org/show_bug.cgi?id=251696</a>
&lt;rdar://104964643&gt;

Reviewed by Simon Fraser.

UI-side scrolling changes need to be reflected on the Interaction
Regions layer side. Otherwise the in-layer coordinates of the regions
will be offset.

We connect ScrollingStateNodes to their respective Interaction Regions
layers and relay all bounds/position changes.

The root interaction regions layer is not impacted as it is parented to
the root node&apos;s main layer.

* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::ScrollingStateNode::setInteractionRegionsLayer):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::interactionRegionsLayer const):
Add an `interactionRegionsLayer` property.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
Set Interaction Regions layers on ScrollingStateNodes.

* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.mm:
(WebCore::ScrollingTreeFixedNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreeFixedNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeOverflowScrollProxyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeOverflowScrollProxyNodeCocoa.mm:
(WebCore::ScrollingTreeOverflowScrollProxyNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreeOverflowScrollProxyNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreePositionedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreePositionedNodeCocoa.mm:
(WebCore::ScrollingTreePositionedNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreePositionedNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
(WebCore::ScrollingTreeStickyNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreeStickyNodeCocoa::applyLayerPositions):
Apply all layer positions changes to the ScrollNode&apos;s corresponding
Interaction Regions Layer.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateBeforeChildren):
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::repositionScrollingLayers):
Apply the ScrollView&apos;s contentOffset to the corresponding Interaction
Regions layer.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::applyCommonPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
(WebKit::applyGeometryPropertiesToLayer): Deleted.
Apply the masksToBounds property to Interaction Regions layers.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::detachFromParent):
Remove the Interaction Regions layer from superlayer on detach even for
UIView based nodes.
(WebKit::RemoteLayerTreeNode::initializeLayer):
Disable implicit position animations for Interaction Regions layers.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):
Improve the debug style for occlusion layers.

Canonical link: <a href="https://commits.webkit.org/260598@main">https://commits.webkit.org/260598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2114dad570657fdb364fc2cf12076a95e1d5700a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17781 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117790 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19232 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9058 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100913 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114447 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14427 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97652 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42404 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96400 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29290 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84138 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10586 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30640 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11346 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7557 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16735 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50236 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12934 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3991 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -177,6 +177,9 @@ static void applyGeometryPropertiesToLayer(CALayer *layer, const RemoteLayerTree
         layer.rasterizationScale = properties.contentsScale;
     }
+
+    if (properties.changedProperties & LayerChange::MasksToBoundsChanged)
```
This is still not a geometry property :)

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -177,6 +177,9 @@ static void applyGeometryPropertiesToLayer(CALayer *layer, const RemoteLayerTree
         layer.rasterizationScale = properties.contentsScale;
     }
+
+    if (properties.changedProperties & LayerChange::MasksToBoundsChanged)
```
Why did this move?

etiennesegonzac replied:
This was always needed (interaction regions need to be clipped too).
But I think the method's name caused confusion :)

hortont424 replied:
Unfortunate that we still can't tell what they are common *between* by the name. But this is less wrong.

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1078,11 +1078,25 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
 }
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION) || ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+- (void)_updateScrollingCoordinator:(const Vector<WebCore::GraphicsLayer::PlatformLayerID>&)destroyedLayers
```
This name is too vague. What does "update" mean here?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
```diff
@@ -71,6 +71,10 @@ class RemoteLayerTreeNode : public CanMakeWeakPtr<RemoteLayerTreeNode> {
     void setStationaryScrollContainerIDs(Vector<WebCore::GraphicsLayer::PlatformLayerID>&& value) { m_stationaryScrollContainerIDs = WTFMove(value); }
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    void reconcileAfterScroll();
```
This doesn't seem to be called "after scroll" as far as I can see. I'd expect a call from `RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidScroll()` or the like, but don't see one.

hortont424 replied:
This seems currently like a very vague name that should perhaps mention interaction regions somehow?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -67,16 +67,29 @@
 }
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+void RemoteLayerTreeNode::reconcileAfterScroll()
```
Can we do this closer to the UIScrollView, like in `ScrollingTreeScrollingNodeDelegateIOS` ?

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
```diff
@@ -65,6 +67,10 @@ class RemoteScrollingCoordinatorProxyIOS final : public RemoteScrollingCoordinat
     void establishLayerTreeScrollingRelations(const RemoteLayerTreeHost&) override;
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    void reconcileConnectedNodes();
```
Same again, this name seems very broad.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -262,8 +300,12 @@
+    else {
         m_uiState.removeNodeWithActiveUserScroll(rootNode->scrollingNodeID());
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+        reconcileConnectedNodes();
```
"reconcile" is too vague. This function is called when a scroll starts, not for every scroll increment, so I'm not sure why fixing up the interaction layer bounds here makes sense?

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
```diff
@@ -29,6 +29,9 @@
 #if PLATFORM(IOS_FAMILY) && ENABLE(ASYNC_SCROLLING)
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+#import "RemoteLayerTreeNode.h"
```
This should move to its own paragraph below (because of the #if)

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -67,16 +67,29 @@
+void RemoteLayerTreeNode::applyScrollUpdatesToInteractionRegionsLayer()
+{
+    CALayer *contentLayer = layer();
+    CALayer *interactionLayer = interactionRegionsLayer();
```
Do we always have one? Should we bail here?

etiennesegonzac replied:
probably safer to check
(looked into lazyly creating the interaction container too, but that's for another branch :))

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -152,7 +152,8 @@ void updateLayersForInteractionRegions(CALayer *layer, RemoteLayerTreeHost& host
                 if (applyBackgroundColorForDebugging) {
-                    [interactionRegionLayer setBackgroundColor:cachedCGColor({ WebCore::SRGBA<float>(1, 0, 0, .1) }).get()];
+                    [interactionRegionLayer setBorderColor:cachedCGColor({ WebCore::SRGBA<float>(1, 0, 0, .2) }).get()];
+                    [interactionRegionLayer setBorderWidth:8];
```
That's a beefy border!

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1078,11 +1078,25 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
 }
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION) || ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+- (void)_removeScrollingNodesFromCoordinator:(const Vector<WebCore::GraphicsLayer::PlatformLayerID>&)destroyedLayers
```
The function name says "nodes" but the arguments are layers. And the underlying data is a set of `GraphicsLayer::PlatformLayerID`.

This would be clearer if:
1. m_fixedScrollingNodeLayerIDs had "interaction region" in the name
2. Don't say "node layerIDs", just "fixed layers"
3. We normally use "viewport constrained" as a term that includes sticky and fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
```diff
@@ -146,6 +146,11 @@ void RemoteScrollingCoordinatorProxy::currentSnapPointIndicesDidChange(WebCore::
 void RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidScroll(ScrollingNodeID scrolledNodeID, const FloatPoint& newScrollPosition, const std::optional<FloatPoint>& layoutViewportOrigin, ScrollingLayerPositionAction scrollingLayerPositionAction)
 {
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    // This is called for all frames (including main).
```
This comment isn't useful here; the scrolling tree crosses frame boundaries.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
```diff
@@ -146,6 +146,11 @@ void RemoteScrollingCoordinatorProxy::currentSnapPointIndicesDidChange(WebCore::
 {
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    // This is called for all frames (including main).
+    scrollingTreeFixedNodesDidUpdate();
```
Why is the scrolledNodeID not passed in, to allow the implementation to be more efficient? Also the name is confusing; we don't know here that a "fixed node did update"; we just know that some scroller scrolled.

etiennesegonzac replied:
Is there an efficient way to go from the scrolledNodeId to the list of the impacted "viewport constrained" layers?
(apart from the HashSet global to the coordinator)

We could also hook into `ScrollingStateFixedNode/ScrollingStateStickyNode` directly, but I'm not sure we want to add RemoteLayerTree concerns to `scrolling/`

smfr replied:
I think hooking up to the scrolling tree is actually the right thing to do here. The scrolling tree knows exactly what layer tree geometry updates are needed when a scroll happens, taking fixed and sticky into account.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -179,6 +179,22 @@
 }
 
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+void RemoteScrollingCoordinatorProxyIOS::scrollingTreeFixedNodesDidUpdate()
```
When `scrollingTreeNodeDidScroll()` (which calls this) is called, we haven't updated all layer positions, so I'm not convinced that it's valid to check the bounds of CALayers here.

etiennesegonzac replied:
Moved it to `applyScrollingTreeLayerPositionsAfterCommit`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -131,8 +145,16 @@
+                if (remoteLayerTreeNode) {
+                    scrollingStateNode.setScrollContainerLayer(remoteLayerTreeNode->layer());
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+                    scrollingStateNode.setInteractionRegionsLayer(remoteLayerTreeNode->interactionRegionsLayer());
```
I don't know that you need to call this here; this is the "scroll container" code path. The call below `currNode->setLayer()` should be sufficient?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -109,19 +109,33 @@
+                if (remoteLayerTreeNode) {
+                    scrollingStateNode.setScrollContainerLayer(remoteLayerTreeNode->layer());
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+                    scrollingStateNode.setInteractionRegionsLayer(remoteLayerTreeNode->interactionRegionsLayer());
```
I don't know that you need to call this here; this is the "scroll container" code path. The call below `currNode->setLayer()` should be sufficient?

etiennesegonzac replied:
double checking, but I think I needed it to sync in `ScrollingTreeScrollingNodeDelegateIOS ::repositionScrollingLayers()` for `overflow:scroll` and iframes (below)

smfr commented on Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
```diff
@@ -173,6 +173,18 @@ void ScrollingStateNode::setLayer(const LayerRepresentation& layerRepresentation
+
+    m_interactionRegionsLayer = layerRepresentation;
+
+    // Depends on other property changed flags.
```
I'd like this to be more specific. In particular, the interactions region is special, in terms of dirty bits, because it doesn't have its own; you're piggybacking on `ScrollingStateNodeProperty::Layer` I think?

etiennesegonzac replied:
and `ScrollContainerLayer` for the `ScrollingTreeScrollingNodeDelegateIOS`

etiennesegonzac committed:
Update the Interaction Regions layer tree after UI-side scrolling
https://bugs.webkit.org/show_bug.cgi?id=251696
<rdar://104964643>

Reviewed by Simon Fraser.

UI-side scrolling changes need to be reflected on the Interaction
Regions layer side. Otherwise the in-layer coordinates of the regions
will be offset.

We connect ScrollingStateNodes to their respective Interaction Regions
layers and relay all bounds/position changes.

The root interaction regions layer is not impacted as it is parented to
the root node's main layer.

* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::ScrollingStateNode::setInteractionRegionsLayer):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::interactionRegionsLayer const):
Add an `interactionRegionsLayer` property.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
Set Interaction Regions layers on ScrollingStateNodes.

* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.mm:
(WebCore::ScrollingTreeFixedNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreeFixedNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeOverflowScrollProxyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeOverflowScrollProxyNodeCocoa.mm:
(WebCore::ScrollingTreeOverflowScrollProxyNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreeOverflowScrollProxyNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreePositionedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreePositionedNodeCocoa.mm:
(WebCore::ScrollingTreePositionedNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreePositionedNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
(WebCore::ScrollingTreeStickyNodeCocoa::commitStateBeforeChildren):
(WebCore::ScrollingTreeStickyNodeCocoa::applyLayerPositions):
Apply all layer positions changes to the ScrollNode's corresponding
Interaction Regions Layer.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateBeforeChildren):
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::repositionScrollingLayers):
Apply the ScrollView's contentOffset to the corresponding Interaction
Regions layer.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::applyCommonPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
(WebKit::applyGeometryPropertiesToLayer): Deleted.
Apply the masksToBounds property to Interaction Regions layers.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::detachFromParent):
Remove the Interaction Regions layer from superlayer on detach even for
UIView based nodes.
(WebKit::RemoteLayerTreeNode::initializeLayer):
Disable implicit position animations for Interaction Regions layers.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):
Improve the debug style for occlusion layers.

Canonical link: https://commits.webkit.org/260598@main
